ERCS/erc-8001.md

```diff
@@ -0,0 +1,283 @@
+    return keccak256(abi.encodePacked("\x19\x01", domainSeparator, _acceptanceStructHash(a)));
+  }
+
+```
+Computation (normative):
+
+```solidity
+participantsHash = keccak256(abi.encodePacked(participants)); // sorted unique
+intentStructHash = keccak256(abi.encode(
+  AGENT_INTENT_TYPEHASH,
+  payloadHash, expiry, nonce, agentId, coordinationType, coordinationValue, participantsHash
+));
+intentDigest = keccak256("\x19\x01" || domainSeparator || intentStructHash);
+
+```
+
+Clarifications (normative):
+- `getIntentHash(intent)` MUST return `intentStructHash` (struct hash), not the full digest.
+- `AcceptanceAttestation.intentHash` MUST be that struct hash.
+- Each acceptance is signed over its own EIP-712 digest that includes this field.
+- `participants` MUST be strictly ascending by uint160(address) and deduplicated.
+
+
+### Interface
+
+Implementations **MUST** expose the following interface and events.
+
+```solidity
+interface IAgentCoordination {
+    event CoordinationProposed(bytes32 indexed intentHash, address indexed proposer, bytes32 coordinationType, uint256 participantCount, uint256 coordinationValue);
+    event CoordinationAccepted(bytes32 indexed intentHash, address indexed participant, bytes32 acceptanceHash, uint256 acceptedCount, uint256 requiredCount);
+    event CoordinationExecuted(bytes32 indexed intentHash, address indexed executor, bool success, uint256 gasUsed, bytes result);
+    event CoordinationCancelled(bytes32 indexed intentHash, address indexed canceller, string reason, uint8 finalStatus);
+
+    function proposeCoordination(AgentIntent calldata intent, bytes calldata signature, CoordinationPayload calldata payload) external returns (bytes32 intentHash);
+    function acceptCoordination(bytes32 intentHash, AcceptanceAttestation calldata attestation) external returns (bool allAccepted);
+    function executeCoordination(bytes32 intentHash, CoordinationPayload calldata payload, bytes calldata executionData) external returns (bool success, bytes memory result);
+    function cancelCoordination(bytes32 intentHash, string calldata reason) external;
+
+    function getCoordinationStatus(bytes32 intentHash) external view returns (Status status, address proposer, address[] memory participants, address[] memory acceptedBy, uint256 expiry);
+    function getRequiredAcceptances(bytes32 intentHash) external view returns (uint256);
+    function getAgentNonce(address agent) external view returns (uint64);
+}
+```
+
+### Semantics
+
+- **Participants** MUST be unique and sorted ascending. Implementations MUST reject non-canonical arrays.
+- `proposeCoordination`:
+    - Verifies EIP-712 signature by `agentId` using ECDSA for EOAs or [ERC-1271] for contracts.
+    - Requires `intent.expiry > block.timestamp` and `intent.nonce > agentNonces[agentId]`.
+    - Stores the canonicalised state and sets `agentNonces[agentId] = intent.nonce`.
```

Comment:
The observable effect of these requirements (having a nonce) is perfectly acceptable, but you'll want to define them in an implementation agnostic way. ERCs shouldn't constrain _how_ a contract implements the interface, only how it behaves.

If, for example, a contract could use something crazy like TSTORE/TLOAD plus a merkle tree to store nonces. That implementation should be equally valid as one using a plain array.

So my suggestion here is to rewrite these requirements so that they describe the expected behaviour, and not a particular implementation.